implementation/src/main/java/io/smallrye/mutiny/operators/multi/MultiConcatMapOp.java

```diff
@@ -0,0 +1,312 @@
+
+        final MultiSubscriber<? super O> downstream;
+        final Function<? super I, ? extends Publisher<? extends O>> mapper;
+        private final boolean delayError;
+
+        final AtomicReference<Throwable> failures = new AtomicReference<>();
+
+        volatile Subscription upstream = null;
+        private static final AtomicReferenceFieldUpdater<ConcatMapMainSubscriber, Subscription> UPSTREAM_UPDATER = AtomicReferenceFieldUpdater
+                .newUpdater(ConcatMapMainSubscriber.class, Subscription.class, "upstream");
+
+        final ConcatMapInner<O> inner;
+
+        ConcatMapMainSubscriber(
+                MultiSubscriber<? super O> downstream,
+                Function<? super I, ? extends Publisher<? extends O>> mapper,
+                boolean delayError) {
+            this.downstream = downstream;
+            this.mapper = mapper;
+            this.delayError = delayError;
+            this.inner = new ConcatMapInner<>(this);
+        }
+
+        // SUBSCRIPTION (called from downstream)
+
+        @Override
+        public void request(long n) {
+            if (n > 0) {
+                if (state.compareAndSet(STATE_NEW, STATE_READY)) {
+                    upstream.request(1);
+                    // No outstanding requests from inner, forward the request to upstream
+                } else if (state.get() == STATE_READY && inner.requested() == 0) {
+                    upstream.request(1);
+                }
+                inner.request(n);
+            } else {
+                downstream.onFailure(new IllegalArgumentException("Invalid requests, must be greater than 0"));
+            }
+        }
+
+        @Override
+        public void cancel() {
+            while (true) {
+                int state = this.state.get();
+                if (state == STATE_CANCELLED) {
+                    return;
+                }
+                if (this.state.compareAndSet(state, STATE_CANCELLED)) {
+                    if (state == STATE_TERMINATED) {
+                        inner.cancel();
+                    } else {
+                        inner.cancel();
+                        upstream.cancel();
+                    }
+                    return;
+                }
+            }
+        }
+
+        // SUBSCRIBER METHODS (called from upstream)
+
+        @Override
+        public void onSubscribe(Subscription s) {
+            if (UPSTREAM_UPDATER.compareAndSet(this, null, s)) {
+                downstream.onSubscribe(this);
+            }
+        }
+
+        @Override
+        public void onItem(I item) {
+            if (!state.compareAndSet(STATE_READY, STATE_EMITTING)) {
+                return;
+            }
+
+            try {
+                Publisher<? extends O> p = mapper.apply(item);
+                Objects.requireNonNull(p, "The mapper returned a null Publisher");
+
+                p.subscribe(inner);
+            } catch (Throwable e) {
```

Comment:
What do we do of the NPE here? This shall terminate the stream with an error.

Reply:
IT must follow the same semantics as the regular flatMap. I believe we failed in this case.

Reply:
So now I do the same as in flatmap: throw an NPE with a specific message.